# Introduction to Programming

# jupyter notebook

This notebook will introduce you to programming in Python, and using the Jupyter Notebook environment.

To begin, Python is what is known as an interpreted language. In other words, the Python program executes your code by interpreting it and translating it to machine language to be executed. This type of language is typically closer to human language, making it easier to learn, but is slower to execute than compiled languages like C and Fortran. Compiled languages takes the code, compiles it into executeble machine code, which can then be run much faster. There are some technologies to improve Python speed (Just in Time compiler), but we will forgo that for now. Speed really isn't an issue yet, we need to first learn how to write code.

The power of Python comes from the large open source community of developers that have done the heavy lifting for many tasks by creating well documented easy to use libraries to help you write useful code quickly. There are particularly many libraries for data analytic functions, which is why Python is often the weapon of choice for data scientists, digital marketers, and data intensive businesses. 

# Download / Setup

Please download Anaconda **with Python 2.7** from [https://www.anaconda.com/download/](https://www.anaconda.com/download/).

Download the appropriate version for your computer.

Once installed, open up command prompt in windows or terminal in OS X.

Launch the jupyter notebook app with the command: `jupyter notebook`.

<img src="./images/terminal.png" alt="terminal" style="width: 400px"/>


You should see a browser pop up launching the address: `loclahost:8888`

<img src="./images/jupyter.png" alt="jupyter" style="width: 400px"/>


Basically, Jupyter opened a local server on the port 8888. This is how you will communicate with your Python.

# General features of the environment

* Think of Jupyter notebooks as a notebook... duh. It is a place where you can play with code and document what you are doing in order to create a cohesive document with code that does stuff.
* Cells. Each bit of code is confied to a cell. Cells can be independently executed. Once a bit of code is executed, a number indicating the order of execution will appear on the left. Cells are selected using the dropdown menu at the top of the notebook. <img src="./images/celltypes.png" alt="jupyter" style="width: 700px"/>
* To execute a cell, you can use the play button, or shift+enter (automatically goes to next cell or insert one if already last cell), or control+enter (executes without going to next cell).
* There are basically 2 types of cells that you will be using. Code cells are the default, and markdown cells for notation. **Double click on this cell to see how markdown works**
* We will get to code cells in just a bit. For now, a quick markdown tutorial.

## Markdown tutorial (`#` indicates headers, the more you put, the smaller the header)
Markdown is a type of markup language that allows you to write a document embedding images, code snippets, math equations, etc. It is super useful!

1. Here's a list
2. Part 2
1. Look, it automatically keeps numbering even though I typed the wrong number.
1. Oops I did it again.

In the lines above the tutorial, I've used markdowns to do a bulleted list, and used HTML code to embed an image.

**Everyone loves math!** (this is how you bold)

For in line math, you can do something like this $y=3x^2 \implies x = \pm\sqrt{\frac{y}{3}}$. Note that math is typed using a flavor of LaTex (via Mathjax). This is a common typesetting language used by scientists to write professional looking equations.

If you want to write a centered math block, you do this:

$$
y=3x^2 \implies x = \pm\sqrt{\frac{y}{3}} \\
\text{you can break lines and do more math! } \sum_{i=1}^N{i}=(N+1)\times\frac{N}{/2}
$$

If you need to figure out how to do someting in LaTex, your best bet is just to google whatever it is you need to do + "latex". Just be careful what word you put before LaTeX.

If you want to embed code, you just do this: `# this is a snippet of code`.

Or, if you want Jupyter to auto highlight syntax for Python:

```Python
print("hello world!" # This code prints "hello world"
```

or Java if you're so inclined...

```Java
public class HelloWorld {
   public static void main(String[] args) {
      // Prints "Hello, World" in the terminal window.
      System.out.println("Hello, World");
   }
}```

For more information about working with Mathjax please see [this cheatsheet](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#videos) $\leftarrow$ this is how you do a hyperlink.

# Let's get programming

## Hello World...right of passage

In [3]:
print('Hello world!') # print(is a default python function that "prints" | "hello world" is just what print(outputs...

Hello world!


### Example
Okay, now that "Hello World" is out of the way, let's first begin with a simple example. Python can do some basic arithmetic.

In [4]:
# comments follow "#" sign
3+2 # we can do addition! 

5

In [5]:
3-2 # we can do subtraction!

1

In [6]:
3*2 # we can do multiplication!

6

In [7]:
3/2 # we can do division... kind of (it works in python 3 as expected)

1

#### What happened here? 
This is a good time to introduce the concept of data types. 3 and 2 are both integers, so the default result of the division of 2 integers is to return another integer. In this case, it is the closest integer lower than the actual result. We can complete the operation as expected by changing one of the data types to "float." A float is a decimal. This means the computer memory will keep track of as many digits as necessary up to a certain number (the machine limit).

In [8]:
3./2 # Look, now it worked! We can do division as expected because of the "." after the 3

1.5

Alternatively, we can import the division from Python 3 that infers the proper data type.

In [9]:
from __future__ import division # this means, from the __future__ module, import the function division
3/2

1.5

In [10]:
5/6 # now all of our division works as expected, as long as the above cell has been executed

0.8333333333333334

### Example 2 - variables

The concept of a variable is essential to programming. A variable is anything that can take on some value. Valid variable names begin with a letter in Python (or an underscore, "_"). Variables are case sensitive.

In [11]:
x = 5 # assign x as 5, note in programming, = typically means set variable as target value
y = 6
z = x+y # assign a variable as a sum of 2 other variables
print(z) # print(is a reserved keyword function in Python that outputs the subsequent code

11


In [12]:
x = 3
z = x+y
print(z)

9


In [13]:
x = 2
print(z)

9


In [14]:
x + q # this won't work because we haven't assigned anything to q

NameError: name 'q' is not defined

In [15]:
q = z
answer = q+z

In [16]:
answer

18

#### Variables don't have to be just numbers, they can also be words, or in programming lingo, "strings"

In [17]:
q = 'this is a string!!!! l33t skill2' #Note that strings are inside quotes. They can be single or double.
print(q)

this is a string!!!! l33t skill2


In [18]:
q + z # you can't add strings with numbers... (remember q is defined as a string and z is defined as the sum of x and y which are integers)

TypeError: cannot concatenate 'str' and 'int' objects

In [19]:
q + ' oh this works!' # You CAN "add" strings with strings, this is known as concatenation

'this is a string!!!! l33t skill2 oh this works!'

In [20]:
q + str(z)

'this is a string!!!! l33t skill29'

In [21]:
print('The answer is: '+str(z))

The answer is: 9


In [22]:
print('The answer is: {} and {}'.format(z, x))

The answer is: 9 and 2


#### Variables can be any data type. Here are some useful native Python data types: lists, sets, dictionaries

A list is a Python data type that is an ordered collection of objects.

In [23]:
list1 = [2, 5, 6, 8, 100, -3.0, -2.1, -10]
print(list1)

[2, 5, 6, 8, 100, -3.0, -2.1, -10]


In [24]:
# Lists have an attribute "length"
print(len(list1)) # len is a default function in Python that computes how long an object is and returns an integer

8


In [25]:
# You can access a specific entry in a list, the entries start at 0.
list1[0]

2

In [26]:
list1[-1] # you can also use negative indices, where -1 is the last entry

-10

In [27]:
list1*3 # you can "multiply" lists, this appends (repeats) the same list to itself 3 times (or whatever you multiply by).

[2,
 5,
 6,
 8,
 100,
 -3.0,
 -2.1,
 -10,
 2,
 5,
 6,
 8,
 100,
 -3.0,
 -2.1,
 -10,
 2,
 5,
 6,
 8,
 100,
 -3.0,
 -2.1,
 -10]

In [28]:
list1+list1 # you can "add" lists

[2, 5, 6, 8, 100, -3.0, -2.1, -10, 2, 5, 6, 8, 100, -3.0, -2.1, -10]

In [29]:
# Lists can also include mixed data types, including other lists... what you put in a list is arbitrary
list2 = [3, 2, 'hello', list1, [x, y], -1]
print(list2)

[3, 2, 'hello', [2, 5, 6, 8, 100, -3.0, -2.1, -10], [2, 6], -1]


In [30]:
list2[4]

[2, 6]

In [31]:
list2[4][1]

6

In [32]:
# You can append lists (append is a functional attribute of a list), note that append replaces the list being appended with the new augmented list
list1.append(list2)
print(list1) # the outtput is too big, so it omits some stuff as denoted by '...'

[2, 5, 6, 8, 100, -3.0, -2.1, -10, [3, 2, 'hello', [...], [2, 6], -1]]


In [33]:
list1[-1][4][0]

2

A set is a Python object that is similar to a list in that it is a collection of entries of objects. However, sets are unordered.

In [34]:
{1, 1, 1, 2}

{1, 2}

In [35]:
set0 = set([1, 1, 1, 2])

In [36]:
set0

{1, 2}

In [37]:
len(set0)

2

In [38]:
set1 = {3,4,9,-1,5, 2} # sets are denoted by curly brackets with entries inside
print(set1)

set([2, 3, 4, 5, 9, -1])


In [39]:
len(set1) #sets also have lengths

6

In [40]:
set1*2 # you can't multiiply sets

TypeError: unsupported operand type(s) for *: 'set' and 'int'

In [41]:
# since sets are unordered, you can't access specific items
set1[0]

TypeError: 'set' object does not support indexing

In [42]:
# you can't add sets
set1+set1

TypeError: unsupported operand type(s) for +: 'set' and 'set'

In [43]:
set1

{-1, 2, 3, 4, 5, 9}

In [44]:
# but you can subtract sets!
set2 = {'a', 3, x, -1}
set1 - set2 # finds only items in set1 but not set 2, note that 5 is in set1 and x, which = 5, is in set 2, is also omitted in this result

{4, 5, 9}

In [45]:
# sets only has unique objects
set3 = {1,200,1,1,1}
print(set3)

set([200, 1])


In [46]:
set1, set2, set3

({-1, 2, 3, 4, 5, 9}, {-1, 2, 3, 'a'}, {1, 200})

In [47]:
# You can get unions of sets, a union is all unique items in both sets.
set4 = set1.union(set2).union(set3) # note you can chain these functions since the result of set1.union(set2) is a set, which also has the union property
# unlike lists append function, you have to specify what to call the resulting union of sets
print(set4)

set(['a', 1, 2, 3, 4, 5, 200, 9, -1])


Why do we use sets? It is very useful for checking membership:

In [48]:
set1

{-1, 2, 3, 4, 5, 9}

In [49]:
# % is an ipythong magic indicator, they're "magic" commands that run only in iPython, not normal python % timeit times the code
%timeit (-1 in set1)

10000000 loops, best of 3: 35.6 ns per loop


In [50]:
list_1 = list(set1) # we can convert the set to a list
print(list_1)

[2, 3, 4, 5, 9, -1]


In [51]:
# it's a little bit slower because the list is small, this becomes a big problem when the list is huge.
%timeit (-1 in list_1)

The slowest run took 12.47 times longer than the fastest. This could mean that an intermediate result is being cached.
10000000 loops, best of 3: 82.5 ns per loop


Lastly, we will look at dictionaries. Dictionaries are objects that map keys to values.

In [53]:
dict1 = {'a':4, 'b':7, 'c':set1} # dictionaries are defined by curly brackets with pair sof keys:values separated by commas
print(dict1)

{'a': 4, 'c': set([2, 3, 4, 5, 9, -1]), 'b': 7}


In [54]:
dict1['c'] # we can look up the values associated with a key in a dictionary by usiing the indexing notation

{-1, 2, 3, 4, 5, 9}

In [55]:
dict1['b']

7

In [56]:
# we can extract both the keys and values from the dictionary
dict1.keys() #.keys is a function

['a', 'c', 'b']

In [57]:
dict1.values() #.values is a function

[4, {-1, 2, 3, 4, 5, 9}, 7]

In [58]:
dict1.items() #we can get a list of tuples, another data type similar to lists, where each tuple is (key, value)

[('a', 4), ('c', {-1, 2, 3, 4, 5, 9}), ('b', 7)]

In [59]:
listx = [4,5,6,7]

In [61]:
listx[-1] = 8
listx

[4, 5, 6, 8]

In [62]:
tuplex = (4,5,6,7)
tuplex

(4, 5, 6, 7)

In [63]:
tuplex[-1]=8 # tuples are immutable

TypeError: 'tuple' object does not support item assignment

# Programming principles: Loops, Booleans, and Functions

A common thing that makes programming powerful is the ability to feed your system a sequence of generic instructions that are flexible enough to be computed over new data. To do this, we need to look at 3 things:

1. Loops - repeating the same tasks for a specified duration or until a condition is met.
1. Booleans - evaluating truth values, allows us to compute different things depending on some conditions
1. Functions - a set of instructions that takes input data and maps on to some output.



## Basic loops

There are basically 2 kinds of loops. The first is to loop for a prespecified number of iterations, this is a "for loop." The second is to loop until some conditions are met, a "while loop."

Let's begin with an example: add the numbers 1... 1000

First we write a for loop.

In [64]:
s = 0 # Start with some s = 0
for i in range(1, 1001): # this line tells Python, for each item, i, in the iterable range 1 to 1001,
    #note this range function constructs a list from 1 to 1000 inclusive. Note the colon to signify an indented block.
    s = s+i # add i to the previous total. Note that the instructions inside the loop are indented one level.
s # this just displays the final value, this is not indented so it is outside of the loop.

500500

Next we write a while loop.

In [ ]:
i = 1 # initiate an i
s = 0 # initiate an s
while i<1001: # note this condition for the while loop
    s = s+i # add to previous sum
    i += 1 # This specifies the next i to add to s. The += 1 notation is the same as i = i+1
s

Every while loop can be written as a for loop and vice versa.

Of course there's a simpler way to do this, we can also use the sum function on lists to do the same thing.

In [65]:
s = sum(range(1,1001)) # sums up the entire list of numbers to add up
s

500500

## Basic Booleans

Many times, we need to evaluate logical statements, these are statements that result in a truth value: True or False.

Herea are some examples

In [66]:
3+2==5 # Note the double equal sign to evaluate equality

True

In [67]:
3+2<5

False

In [68]:
3+2>5

False

In [69]:
3+2<=5 # Note <= is less than or equal to

True

In [70]:
(3+2==5)&(3+2>5) # you can combine booleans, this is an "and" condition.

False

In [71]:
(3+2==5)|(3+2>5) # This is an or condition

True

In [75]:
(3+2==5)|((3+2>5)&(3+1>5))

True

We can insert booleans into loops

In [76]:
s = 0
for i in range(1,1001):
    if i % 2 == 0: # This evaluates the remainder (or modulo) function
        s = s+i # note a second indent, since this bit of code is evaluated when the if condition is satisfied
s # this is sum of all the even numbers from 1 to 1000

250500

In [77]:
s = 0
for i in range(1,1001):
    if i % 2 != 0: # This evaluates the remainder (or modulo) function
        s = s+i # note a second indent, since this bit of code is evaluated when the if condition is satisfied
s # this is the sum of all the odd numbers from 1 to 1000

250000

Python allows us to write these simple loops more compactly using list comprehension

In [80]:
sum([v for v in range(1,1001) if (v % 2 == 0)]) #sum all divisible by 2

250500

In [79]:
sum([v for v in range(1,1001) if (v % 2 != 0)]) #sum all divisible by 2

250000

In [82]:
sum([v for v in range(1,1001) if (v % 3 == 0)&(v%2==0)]) #sum all divisible by 2

83166

## Functions
$f(x)=$do something with x and return the result

Let's write a function that takes a beginning, $n$, and an end value, $m$, to evaluate a consecutive sum.


In [83]:
def add_consecutive(n,m): # This defines a function that takes n and m as inputs, also see the colon for indented block
    s = 0 
    for i in range(n, m+1): # note where n and m are being used inside the function.
        s = s+i
    return s

So... what happened?

In [84]:
add_consecutive # we've defined a function that's now in the memory

<function __main__.add_consecutive>

In [88]:
s = add_consecutive(0,99999) # we need to give it values for n and m to execute.

In [90]:
s = add_consecutive(7,3000)

Python allows functions to have both ordered inputs (like the one above, where order matters) and named inputs. Let's write a function that allows us to add every $k^{\text{th}}$ number in a list.

In [91]:
def add_consecutive_k(n,m, k=1): # We initialize the kth input as 1, so that if it is not explicitly specified, 
    # the function assumes k=1
    s = 0 
    for i,j in enumerate(range(n, m+1)): # enumerate takes the list and returns the order, value pair
        # here that pair is i (order) and j (value)
        if (i+1)%k==0: # See how we used i and k here. Remember that Python is zero indexed. So the first item in an 
            # iterable is actually 0. THe first item in the range n to m is the 0th item. 
            s = s+j # note we add j to s not i, since j corresponds to the value in the list.
    return s

In [92]:
add_consecutive(0,99778), add_consecutive_k(0,99778), add_consecutive_k(0,99778, k=1) # we get the same answers, yay!

(4977874531L, 4977874531L, 4977874531L)

In [96]:
add_consecutive_k(0,1000) # now we can specify to only add every 3rd number. 

500500

In [97]:
k

NameError: name 'k' is not defined

# Okay, we said that Python was super useful because it has lots of libraries. Let's look at one useful one, Pandas.

Pandas is a dataframe library that allows us to read in / save data and manipulate it in a tabular format.

In [99]:
import pandas as pd # we are importing the entire module, and giving it a nickname "pd", conda installs this library by default
# If you find a package that may be useful, it will generally tell you how to install it. Typically it involves opening command prompt in windows
#or console in Mac and typing conda install package_name or pip install package_name 

#### Pandas has tons of functions and capaility to read lots of file formats. One of the most common is csv (or comma separated value)

In [100]:
# let's read in an example csv file
df = pd.read_csv('https://www.dropbox.com/s/4bbkolouwwrt0dj/data_sales_intro_programming.csv?dl=1', sep = ',') 
# this function takes a bunch of options, one of them is the separator, sep =, it specifies comma by default.
# however, depending on how the file structured, you can specify an appropriate separator
print('This dataset has ' + str(len(df)) + ' entries.')

This dataset has 181 entries.


In [102]:
df.head()

date  sales  ad
0  1/1/2017    857   0
1  1/2/2017    618   0
2  1/3/2017    762   0
3  1/4/2017    628   0
4  1/5/2017    849   1

In [103]:
df.columns.values

array(['date', 'sales', 'ad'], dtype=object)

In [105]:
# df.head() # .head() let's you see the first few rows of data

In [106]:
df[50:60]

date  sales  ad
50  2/20/2017    885   1
51  2/21/2017    899   0
52  2/22/2017    691   1
53  2/23/2017    901   1
54  2/24/2017   1821   1
55  2/25/2017   1476   1
56  2/26/2017    513   0
57  2/27/2017    798   0
58  2/28/2017    664   0
59   3/1/2017    847   0

Note this is a dataset with 3 columns and 181 entries. Each entry consists of a date, the sales total, and whether or not an ad was shown that morning.

In [107]:
df['date'][0] # we can index a dataframe in several methosd

'1/1/2017'

In [108]:
df.date[0]

'1/1/2017'

In [109]:
df.ix[0]

C:\Users\yangy\AppData\Local\conda\conda\envs\py27\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


date     1/1/2017
sales         857
ad              0
Name: 0, dtype: object

In [110]:
df.loc[0:10, ['sales', 'date']]

sales       date
0     857   1/1/2017
1     618   1/2/2017
2     762   1/3/2017
3     628   1/4/2017
4     849   1/5/2017
5    1670   1/6/2017
6    1298   1/7/2017
7     778   1/8/2017
8     874   1/9/2017
9     510  1/10/2017
10   1164  1/11/2017

In [111]:
type(df.date[0]) # It looks like the data we read in is fine,
#but if we look at the data type for one of the date entries, we see that it is a string an not a date
# type is a native Python function to tell you what is the data type

str

In [112]:
df.date = pd.to_datetime(df.date) # we can convert the dates using pandas' to_datetime function

In [113]:
type(df.date[0]) # Now the data type is pandas' datetime index

pandas._libs.tslib.Timestamp

In [114]:
df.loc[0, 'date']

Timestamp('2017-01-01 00:00:00')

In [115]:
df.head()

date  sales  ad
0 2017-01-01    857   0
1 2017-01-02    618   0
2 2017-01-03    762   0
3 2017-01-04    628   0
4 2017-01-05    849   1

In [117]:
# Converting to pandas timestamp allows us to use datetime modules
import datetime # date library
df['dow'] = df.date.map(datetime.date.weekday) #the weekday function converts day of week from 0 to 6, where 0 is Monday and 6 is Sunday
#  df.date is a pandas series (column), .map is an attribute of pandas dataframes/series that maps a function to each value in series

In [118]:
df.head()

date  sales  ad  dow
0 2017-01-01    857   0    6
1 2017-01-02    618   0    0
2 2017-01-03    762   0    1
3 2017-01-04    628   0    2
4 2017-01-05    849   1    3

Let's convert the 0-6 to words, to do so we first define a dictionary. First, we create a list of keys (number dow format) and a list of values (text dow format). We then "zip" them together into a dictionary.

In [120]:
numdow = range(0,7) # range creates a consecutive list of integers starting with the first entry and ending with the last entry - 1 (not incusive)
textdow = ['Mon','Tue','Wed','Thu','Fri','Sat', 'Sun']
dowdict = dict(zip(numdow, textdow)) # zip creates a "zip" object that zips 2 lists together, we convert this zipped list to a dictionary

In [121]:
dowdict

{0: 'Mon', 1: 'Tue', 2: 'Wed', 3: 'Thu', 4: 'Fri', 5: 'Sat', 6: 'Sun'}

In [122]:
dowdict[6] # we can access the word associated with each entry

'Sun'

In [123]:
df.dow = df.dow.apply(lambda x: dowdict[x]) # lambda is a temporary "function" that acts on x, 
# Dataframe allows us to also apply a function to each entry of a column
# If each entry is x, then our temporary function of dowdict[x] will return the corresponding string format dow to the existing
# numerical entry for dow
# altogether, this syntax says, on df.dow apply the lambda function of evaluating dowdict for each value x of df.dow

In [124]:
df.head()

date  sales  ad  dow
0 2017-01-01    857   0  Sun
1 2017-01-02    618   0  Mon
2 2017-01-03    762   0  Tue
3 2017-01-04    628   0  Wed
4 2017-01-05    849   1  Thu

In [ ]:
# we can also convert the dates to month
df['ym'] = df.date.apply(lambda x: datetime.date(x.year, x.month, 1)) #datetime function takes 3 ordered arguments (year, mth, day)

In [ ]:
df.head() # now we can see that dow is comprised of strings, and we have the YM

#### One awesome thing about dataframes is that we can group by variables to do conditional summaries

In [ ]:
# let's look at the average sales by dow
df.groupby(['dow', 'ad'])[['sales']].mean() # we first groupby a variable (dow), then select a remaning column (sales) and apply the 
# summary function "mean" to calculate the conditional mean of sales by day of week
# When are sales the highest?

#### Can we look at conditional averages of sales on both dow and advertising?

In [ ]:
df_summary = df.groupby(['dow', 'ad'])['sales'].mean().unstack('ad')
df_summary

#### What's the effect of advertising?

In [ ]:
df_summary['ad_diff'] = df_summary[1]-df_summary[0]
df_summary

##### Is there a difference by month?

In [ ]:
df.groupby(['ym', 'ad'])['sales'].mean().unstack('ad')

## These tables are all fine and good, can we visualize it?

In [ ]:
# to do so, we should import some more libraries %matplotlib allows the notebook to display plots
% matplotlib inline
from matplotlib import pyplot as plt

First let's plot the time series at the daily level

In [ ]:
plt.figure(figsize=(12,8))
plt.plot(df.date, df.sales, color = 'purple')

Next, let's look at the time series seprately for ad and no ad days.

In [ ]:
plt.plot(df.loc[df.ad == 1, 'date'], df.loc[df.ad == 1, 'sales'])
plt.plot(df.loc[df.ad == 0, 'date'], df.loc[df.ad == 0, 'sales'])

We can do the same plot using only pandas. But Pandas does not "fill in the missing values"

In [ ]:
df.groupby(['date','ad'])['sales'].mean().unstack('ad')

In [ ]:
df.groupby(['date','ad'])['sales'].mean().unstack('ad').plot(figsize=(8,6))

Let's look at the ad vs. no ad by day of week

In [ ]:
df.groupby(['dow', 'ad'])['sales'].mean().unstack('ad').sort_values(by = 0)

In [ ]:
df.groupby(['dow', 'ad'])['sales'].mean().unstack('ad').sort_values(0).plot(kind='bar', colors = ['green', 'red'], figsize=[8,6]) # we also sort the table on values for no ad days

# Let's say your boss wants to see the graph above for each month

### How do you do this?

To do this effectively, we ought to introduce the concept of loops. There are 2 types: for loops and while loops. 

For loops execute some code "for" some list of values.

While loops execute "while" some condition is satisfied.

## So which type of loop should we use to solve the problem posed by your boss?

The for loop is the most natural, we know exactly which values to loop over (months)

In [ ]:
mths = df.ym.unique()
mths

In [ ]:
mths = df.ym.unique() # first we grab the unique months in our dataset
for m in mths:
    df[df.ym == m].groupby(['dow', 'ad'])['sales'].mean().unstack('ad').sort_values(by=0).plot(kind='bar', title = str(m))

## Putting it all together!
### What if we want to be able to streamline the whole process from reading a file to plotting so we can do this task easily in the future?

Let's write a function that takes the file path as an input and saves the file in some output path.

In [ ]:
def read_file_make_mthly_plot(infile, outpath): # note that the folder of outpath must already exist!
    df = pd.read_csv(infile, sep = ',') # read in file
    df.date = pd.to_datetime(df.date) # convert date to proper data type
    df['dow'] = df.date.map(datetime.date.weekday) # create dow variable
    numdow = range(0,7)
    textdow = ['Mon','Tue','Wed','Thu','Fri','Sat', 'Sun']
    dowdict = dict(zip(numdow, textdow)) # create conversion dictionary to text format
    df.dow = df.dow.apply(lambda x: dowdict[x]) # convert dow data to text format
    df['ym'] = df.date.apply(lambda x: datetime.date(x.year, x.month, 1)) # create ym variable
    # next we loop through each month and plot the sales split on advertising
    mths = df.ym.unique() # first we grab the unique months in our dataset
    out = list() # we will collect where the files are saved in this out list, first we define an empty list
    for m in mths:
        df[df.ym == m].groupby(['dow', 'ad'])['sales'].mean().unstack('ad').sort_values(by=0).plot(kind='bar', title = str(m), figsize = (15,10))
        fname = outpath + str(m.year)+'_'+str(m.month)+'.png'
        plt.savefig(fname, dpi = 230, bbox_inches='tight') # this saves the figure with a specified file name = fname and the dpi argument specifies the resolution 
        plt.close() # close the plot so it won't be displayed in the notebook
        out.append(fname) # we append out with newest file
    return out

In [ ]:
out = read_file_make_mthly_plot('https://www.dropbox.com/s/4bbkolouwwrt0dj/data_sales_intro_programming.csv?dl=1', './figures/')
print(out) # The '.' means the current directory

## Voila!
Now your monthly task has gotten a lot easier!